File: abc254/e2.cpp
// 最短距離という意味でBFSの方が解きやすいかも
#include <bits/stdc++.h>
#define ll long long
using namespace std;

int main()
{
    int N, M;
    vector<vector<ll>> graph;
    vector<ll> visited;
    cin >> N >> M;
    graph.assign(N, vector<ll>());
    for (int i = 0; i < M; i++)
    {
        int a, b;
        cin >> a >> b;
        a--;
        b--;
        graph[a].push_back(b);
        graph[b].push_back(a);
    }
    int Q;
    cin >> Q;
    vector<ll> ans(Q);
    for (int i = 0; i < Q; i++)
    {
        int x, k;
        cin >> x >> k;
        x--;
        //
        queue<ll> q;
        map<ll, ll> distance_from_x;
        distance_from_x[x] = 0;
        vector<ll> connected;
        q.push(x);
        while (q.size() != 0)
        {
            auto cur = q.front();
            q.pop();
            connected.push_back(cur);
            if (distance_from_x[cur] == k)
            {
                continue;
            }
            for (auto next : graph[cur])
            {
                if (distance_from_x.count(next) == 0)
                { // not visited
                    q.push(next);
                    distance_from_x[next] = distance_from_x[cur] + 1;
                }
            }
        }
        for (auto c : connected)
        {
            ans[i] += (c + 1);
        }
    }
    for (auto a : ans)
    {
        cout << a << endl;
    }
}

File: abc254/e2.py
from collections import deque
N,M = map(int,input().split())
graph = [[] for i in range(N)]

for i in range(M):
    a,b = map(int,input().split())
    a-=1
    b-=1
    graph[a].append(b)
    graph[b].append(a)


Q = int(input())
ans_ls = []
for i in range(Q):
    x,k = map(int,input().split())
    x -= 1
    # BFS
    cost_to_vertices = {}
    cost_to_vertices[x] = 0
    queue = deque([x])

    while len(queue) > 0:
        cur = queue.popleft()
        if cost_to_vertices[cur] == k:
            continue
        
        for n in graph[cur]:
            if not (n in cost_to_vertices):
                cost_to_vertices[n] = cost_to_vertices[cur] + 1
                queue.append(n)
    ans = 0
    for k,v in cost_to_vertices.items():
        ans += (k+1)
    ans_ls.append(ans) 

for a in ans_ls:
    print(a)